services/ui_backend_service/api/ws.py

```diff
@@ -0,0 +1,226 @@
+        if since:
+            # Subtract 1 second to make sure all events are included
+            event_queue = await self.queue.values_since(since)
+            for _, event in event_queue:
+                self.loop.create_task(
+                    await self._event_subscription(subscription, event['operation'], event['resources'], event['data'])
+                )
+
+    async def unsubscribe_from(self, ws, uuid: str = None):
+        if uuid:
+            self.subscriptions = list(
+                filter(lambda s: uuid != s.uuid or ws != s.ws, self.subscriptions))
+        else:
+            self.subscriptions = list(
+                filter(lambda s: ws != s.ws, self.subscriptions))
+
+    async def handle_disconnect(self, ws):
+        """
+        Sets disconnected timestamp on websocket subscription without removing it from the list.
+        Removing is handled by event_handler that checks for expired subscriptions before emitting
+        """
+        self.subscriptions = list(
+            map(
+                lambda sub: sub._replace(disconnected_ts=time.time()) if sub.ws == ws else sub,
+                self.subscriptions)
+        )
+
+    async def websocket_handler(self, request):
+        "Handler for received messages from the open Web Socket connection."
+        # TODO: Consider using options autoping=True and heartbeat=20 if supported by clients.
+        ws = web.WebSocketResponse()
+        await ws.prepare(request)
+
+        while not ws.closed:
+            async for msg in ws:
+                if msg.type == WSMsgType.TEXT:
+                    try:
+                        # Custom ping message handling.
+                        # If someone is pinging, lets answer with pong rightaway.
+                        if msg.data == "__ping__":
+                            await ws.send_str("__pong__")
+                        else:
+                            payload = json.loads(msg.data)
+                            op_type = payload.get("type")
+                            resource = payload.get("resource")
+                            uuid = payload.get("uuid")
+                            since = payload.get("since")
+                            if since is not None and str(since).isnumeric():
+                                since = int(since)
+                            else:
+                                since = None
+
+                            if op_type == SUBSCRIBE and uuid and resource:
+                                await self.subscribe_to(ws, uuid, resource, since)
+                            elif op_type == UNSUBSCRIBE and uuid:
+                                await self.unsubscribe_from(ws, uuid)
+                    except Exception:
+                        self.logger.exception("Exception occurred.")
+
+        # Always remove clients from listeners
+        await self.handle_disconnect(ws)
+        return ws
+
+    @throttle_simultaneous(count=8)
```

Comment:
this could _maybe_ be replaced with BoundedSemaphore from the asyncio library. Would get rid of a possibly unnecessary dependency